cmd/oras/attach.go

```diff
@@ -0,0 +1,133 @@
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/file"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/option"
+	"oras.land/oras/cmd/oras/internal/parse"
+	"oras.land/oras/internal/content"
+)
+
+type attachOptions struct {
+	option.Common
+	option.Remote
+	option.Pusher
+
+	targetRef    string
+	artifactType string
+	fileRefs     []string
+}
+
+func attachCmd() *cobra.Command {
+	var opts attachOptions
+	cmd := &cobra.Command{
+		Use:   "attach name[:tag|@digest] file[:type] [file...]",
+		Short: "Attach files to an existed manifest",
+		Long: `Attach files to an existed manifest
+
+Example - Attach file 'hi.txt' with type 'sig/example' to manifest 'hello:test' in registry 'localhost:5000'
+  oras attach localhost:5000/hello:test hi.txt --artifact-type sig/example
+`,
+		Args: cobra.MinimumNArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return opts.ReadPassword()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			opts.fileRefs = args[1:]
+			return runAttach(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.artifactType, "artifact-type", "", "", "artifact type")
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runAttach(opts attachOptions) error {
+	ctx, _ := opts.SetLoggerLevel()
+
+	// Prepare manifest
+	store := file.New("")
+	defer store.Close()
+	dst, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+	if dst.Reference.Reference == "" {
+		return newErrInvalidReference(dst.Reference)
+	}
+	subject, err := dst.Resolve(ctx, dst.Reference.Reference)
+	if err != nil {
+		return err
+	}
+	var refs []content.FileReference
+	for _, ref := range opts.fileRefs {
+		refs = append(refs, content.NewFileReference(parse.FileReference(ref, "")))
+	}
+	ociDescs, err := content.LoadFiles(ctx, store, nil, refs, opts.Verbose)
+	if err != nil {
+		return err
+	}
+	orasDescs := make([]artifactspec.Descriptor, len(ociDescs))
+	for i := range ociDescs {
+		orasDescs[i] = *OCIToArtifact(ociDescs[i])
+	}
+	manifestDesc, err := oras.PackArtifact(
+		ctx, store, opts.artifactType, orasDescs,
+		oras.PackArtifactOptions{
+			Subject: OCIToArtifact(subject),
+		})
+	if err != nil {
+		return err
+	}
+
+	// Prepare Push
+	copyOptions := oras.DefaultCopyOptions
+	copyOptions.PreCopy = display.PreCopyStatus(func() bool { return !opts.Verbose })
+	copyOptions.OnCopySkipped = display.CopySkippedStatus
+
+	// Push
+	err = oras.CopyGraph(ctx, store, dst, manifestDesc, copyOptions.CopyGraphOptions)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Files attached to", opts.targetRef)
+	fmt.Println("Digest:", manifestDesc.Digest)
```

Comment:
This message might be confusing. Digest of what?

Reply:
Good catch! Specified `ORAS Artifact`